### PR TITLE
Dltp 769 generate and schedule periodic 

### PR DESCRIPTION
**Pull request by banurekha**
## Claiming issue DLTP-769

@b59b261f0b104fc594a1b7ba02ddd781621ce71c

$ ./script/start-issue DLTP-769

[skip ci]

## reworking on generic_file extract query

@b479e75b2cb2a19d86d5213a4bd6ed6f817fc529

Rewriting the query and how the values are accessed

## Custom byte to GB conversion

@e197463e3404b082a7edbb6eea1020711f6669a5

## Appeasing Rubocop and increase coverage

@5d8b433ed289785f6b8b265ea9f7f210bc72c327

## Adding script to run metrics report weekly

@ba4049f52bfbbda21440035649b6a94a96d73be3

## Adding script for aggregation_keys migration

@45df00a26db79ede92b2d5aa6591cc2a50798959

## Remove empty elements from descMetadata

@b1ee87d3666c03164449970881fcea9d27e4dc8d

parse_triples method return empty element causing to store blank values in database. This change will remove the empty elements from returned array.

## Store report to database and email report to given list of email

@f40c329c24a11e049d1b8fb50f6aad5d6543f5a8

## Reworking on email part and adding spec

@97121e63e79cb203a0160e34e983a8178d3cf76f

Adding spec for email section and code changes related to that

## Fixing spec failure

@09b3ac6222cf4c317f7708641cd999e70856767f

## Increasing coverage

@be1c41e8d813b3188da3bd0c005b60f20edb63f0

## Moving smtp_config infomation from smtp_config.yml to application.yml

@f93f797fc7814ab53e877abb8faf6771bd2b942b

## Fixing Date format error in shell script

@91fde90775a44a211bfbf4c261cf408a1e05399e

Change date format in shell script. Removing filename on saving and adding index to reporting table

## Error accessing sender email

@61fab088fb931273be04ed830ebd486f09c8a54e

Missed to access default sender from ENV instead on smtp_config.yml. So fixed that and related code and spec
q

### CODE REVIEW

File: app/mailers/report_mailer.rb [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+  private
+
+  def subject(report)
+    "CurateND Metrics Report: #{report.start_date} Through #{report.end_date} - [#{Rails.env}]"
+  end
+
+  def recipients_list
+    @list ||= Array.wrap(ENV.fetch('METRICS_REPORT_RECIPIENT'))
+    @list
```
Comment (jeremyf) [inline app/mailers/report_mailer.rb]: Line 19 is not needed; the `@list ||=…` line will return the `@list`.

File: app/services/metrics_report.rb [hunk 2]
```diff
@@ -203,8 +218,16 @@ def render
...
-    end
+    report = PeriodicMetricReport.find_or_initialize_by(start_date: metrics.report_start_date,
+                                                        end_date: metrics.report_end_date)
+    report.update!(
+      content: render
+    )
+    send_report(report.id)
+  end
+
+  def send_report(report_id)
```
Comment (jeremyf) [inline app/services/metrics_report.rb]: Is there a reason for using the `report.id`? You should be able to pass the `report` object from the above `save!` method. You already have the object, but in passing the `report.id` you need to requery the database.

File: app/mailers/report_mailer.rb [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+  end
+
+  private
+
+  def subject(report)
+    "CurateND Metrics Report: #{report.start_date} Through #{report.end_date} - [#{Rails.env}]"
+  end
+
+  def recipients_list
+    @list ||= Array.wrap(ENV.fetch('METRICS_REPORT_RECIPIENT'))
```
Comment (jeremyf) [inline app/mailers/report_mailer.rb]: Use `Figaro.env.METRICS_REPORT_RECIPIENT` to retrieve this key. You will need to split the given environment variable on commas (as all ENV variables are strings). The instance variable name should be the same as the method name.

```ruby
@recipient_list ||= Figaro.env.METRICS_REPORT_RECIPIENT!.splti(',').map(&:strip)
```

File: app/mailers/report_mailer.rb [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    @list
+  end
+
+  def default_sender
+    ENV.fetch('METRICS_REPORT_SENDER')
```
Comment (jeremyf) [inline app/mailers/report_mailer.rb]: Use Figaro to fetch this variable.

**Comment by dbrower** [comment]
I received a metrics report email this morning! Very cool.

The title of the email ends with the environment name in brackets, e.g. `[staging]`. Could you move that to the beginning of the email title for non-production environments? It would make it easier for me to sort through them in my inbox.

**Comment by dbrower** [comment]
One other small change...in the heading of the report, also give the environment it is for, e.g.

> CurateND Metrics Report [staging]
> CurateND Metrics Report [pre_production]
> CurateND Metrics Report [production]

We can deal with the many staging machines in some other way.

I foresee getting a metric ton of these email reports.

**Pull request merged**